proto/buf/alpha/module/v1alpha1/module.proto

```diff
@@ -63,4 +63,6 @@ message ModulePin {
   string commit = 5;
   string digest = 6;
   google.protobuf.Timestamp create_time = 7;
+  bool deprecated = 8;
+  string deprecation_message = 9;
```

Comment:
As discussed, including this information on the `ModulePin` is necessary for letting the user know they've updated their dependencies to a deprecated module. And as you've mentioned before, it might be necessary that we also include this information on the `Module` so that the same could be done for other commands like `buf build` (aka anyone that relies on the `DownloadService`)?


If we're worried about including this information on the `Module[Pin]` itself (since it's only used to emit warnings for particular RPCs), we _could_ include it alongside the `ModulePins` returned in the `GetModulePinsResponse`.

In the following examples, I'm referencing `ModuleState` as the `deprecated[_message]` values you have above so that it could eventually support a retracted state, but I'm not even sure that'd be necessary since we might not ever need to return whether or not a module is retracted - we could just return an error on the server-side.

```protobuf
message GetModulePinsResponse {
  repeated buf.alpha.module.v1alpha1.ModulePin module_pins = 1;
  repeated buf.alpha.module.v1alpha1.ModuleState module_states = 2;
}
```

Similarly, we could do this for the `DownloadResponse`, too:

```protobuf
message DownloadResponse {
  buf.alpha.module.v1alpha1.Module module = 1;
  buf.alpha.module.v1alpha1.ModuleState module_state = 2;
}
```

But the only thing that gains us is not tying the idea of a `deprecated[_message]` to the `Module[Pin]` itself, and it actually makes the API less ergonomic and easier to mess up. However, it prevents any changes to the primary module definitions, so it might be worth considering.

Reply:
Thinking about this more - 

It might be that we end up just having a `deprecation_message` field on each of `Module`, `ModulePin` and `Repository` and call it there. We probably don’t need to go overboard and support a generic state like `Retracted` since we can just return server-side errors for those states if we ever need to (e.g. if I try to download a retracted module, we just return an error, not the `Module` type).

Even though the `ModulePin` has a `deprecation_message` field, we don't have to serialize it in the `buf.lock`. And if we ever want to support a `deprecated: <string>` configuration in the `buf.yaml`, we'll already have an avenue to do so since the `Module` will already have it.

Reply:
Discussed offline - 

Let's lean into deprecation being part of the `Repository` and start by determining whether or not a `Module[Pin]` is deprecated by interacting with the `RepositoryService`. It will require an additional round-trip for the commands that care about deprecation, but it’s easier to start with this and change it later if need be.

Reply:
I think we want to remove `deprecated[_message]` from the `ModulePin` type now that we have plans for `GetRepositoriesByFullName``